File: combination-sum/combo_sum.go
package main

import "fmt"

func combinationSum(candidates []int, target int) [][]int {
	var acc []int
	return iter(candidates, target, acc, 0)
}

func iter(candidates []int, target int, acc []int, idx int) [][]int {
	var cs [][]int

	for i := idx; i < len(candidates); i++ {
		t := target - candidates[i]
		if t < 0 {
			continue
		}

		acc = append(acc, candidates[i])
		if t == 0 {
			b := make([]int, len(acc))
			copy(b, acc)
			cs = append(cs, b)
			continue
		} else {
			r := iter(candidates, t, acc, i)
			cs = append(cs, r...)
		}
		acc = acc[:len(acc)-1]
	}

	return cs
}

func main() {
	test1c := []int{2, 3, 6, 7}
	test1t := 7
	a1 := combinationSum(test1c, test1t)
	fmt.Printf("Test1: %v\n", a1)

	test2c := []int{2, 3, 5}
	test2t := 8
	a2 := combinationSum(test2c, test2t)
	fmt.Printf("Test2: %v\n", a2)

	test3c := []int{7, 3, 2}
	test3t := 18
	a3 := combinationSum(test3c, test3t)
	fmt.Printf("Test3: %v\n", a3)
}

File: combination-sum/combo_sum.js
function comboSum(a, t) {
		acc = [];
		comboSumRec(a, 0, t, [], acc);
		return acc;
}

function comboSumRec(a, idx, t, l, acc) {
		if(idx >= a.length) {
				return;
		}

		for(let i=idx; i<a.length; i++) {
				const target = t-a[i];

				// this doesn't work, move on
				if(target < 0) {
						continue;
				} else if (target == 0) {
						// found it
						l.push(a[i]);
						acc.push([...l]);
						l.pop();
				} else {
						// getting closer, continue to look for it
						l.push(a[i]);
						comboSumRec(a, i, target, l, acc);
						l.pop();
				}
		}
}

const a1 = [2, 3, 6, 7];
const t1 = 7;
const test1 = comboSum(a1, t1);
console.log("Test1:",test1);

const a2 = [2, 3, 5];
const t2 = 8;
const test2 = comboSum(a2, t2);
console.log("Test2:",test2);

const a3 = [7, 3, 2];
const t3 = 18;
const test3 = comboSum(a3, t3);
console.log("Test3:",test3);
